azure/services/virtualmachines/virtualmachines.go

```diff
@@ -111,6 +114,34 @@ func (s *Service) Reconcile(ctx context.Context) error {
 		}
 		s.Scope.SetAddresses(addresses)
 		s.Scope.SetVMState(infraVM.State)
+
+		spec, ok := vmSpec.(*VMSpec)
+		if !ok {
+			return errors.Errorf("%T is not a valid VM spec", vmSpec)
+		}
+		idsClient := identities.NewClient(s.Scope)
+
+		// Check if any userAssignedIdentities are missing from the VM and set a condition if so
+		if len(spec.UserAssignedIdentities) == len(infraVM.UserAssignedIdentities) {
+			for _, expectedIdentity := range spec.UserAssignedIdentities {
+				expectedClientID, err := idsClient.GetClientID(ctx, expectedIdentity.ProviderID)
+				if err != nil {
+					return errors.Wrap(err, "failed to get client ID")
+				}
+				found := false
+				for _, actualIdentity := range infraVM.UserAssignedIdentities {
+					if expectedClientID == actualIdentity.ProviderID {
+						found = true
+						break
+					}
+				}
+				if !found {
+					s.Scope.SetConditionFalse(infrav1.VMRunningCondition, infrav1.VMUnhealthyReason, clusterv1.ConditionSeverityWarning, "VM is missing expected user assigned identity")
```

Comment:
What would you think about including the identity name or resource ID in the message to help the user know which one is missing?

Reply:
I think that would be great! I do have a check for the length of the two lists, and if they aren't equal, I immediately mark the condition as fault. However, in this case, I don't know which identity is missing. Should I remove this check and favor knowing which identity is missing? Or should we prioritize the efficiency here.

Reply:
Here's the [line](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/5c806269df45b9a25682784ae275fef46088874a/azure/services/virtualmachines/virtualmachines.go#L142) I'm talking about

Reply:
Since we don't expect more than a few identities per VM I think I'm leaning towards optimizing for UX here and always printing the name of the missing one. Maybe this is also good motivation to optimize the for loops as you pointed out above https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2743#discussion_r1002207518